service/history/replication/task_ack_manager.go

```diff
@@ -97,42 +98,73 @@ func NewTaskAckManager(
+type getTasksResult struct {
+	previousReadTaskID int64
+	lastReadTaskID     int64
+	msgs               *types.ReplicationMessages
+	taskInfos          []*persistence.ReplicationTaskInfo
+	isShrunk           bool
 }
 
-func (t *TaskAckManager) GetTasks(ctx context.Context, pollingCluster string, lastReadTaskID int64) (*types.ReplicationMessages, error) {
+func (t *TaskAckManager) getTasks(ctx context.Context, pollingCluster string, lastReadTaskID int64) (*getTasksResult, error) {
+	var (
+		oldestUnprocessedTaskTimestamp = t.timeSource.Now().UnixNano()
+		oldestUnprocessedTaskID        = t.ackLevels.GetTransferMaxReadLevel()
+		previousReadTaskID             = t.ackLevels.GetClusterReplicationLevel(pollingCluster)
+	)
+
 	if lastReadTaskID == common.EmptyMessageID {
-		lastReadTaskID = t.ackLevels.GetClusterReplicationLevel(pollingCluster)
+		lastReadTaskID = previousReadTaskID
 	}
 
 	taskGeneratedTimer := t.scope.StartTimer(metrics.TaskLatency)
+	defer taskGeneratedTimer.Stop()
+
+	batchSize := t.dynamicTaskBatchSizer.value()
+	t.scope.UpdateGauge(metrics.ReplicationTasksBatchSize, float64(batchSize))
 
-	tasks, hasMore, err := t.reader.Read(ctx, lastReadTaskID, t.ackLevels.GetTransferMaxReadLevel())
+	taskInfos, hasMore, err := t.reader.Read(ctx, lastReadTaskID, t.ackLevels.GetTransferMaxReadLevel(), batchSize)
 	if err != nil {
 		return nil, err
 	}
-	t.scope.RecordTimer(metrics.ReplicationTasksFetched, time.Duration(len(tasks)))
+	t.scope.RecordTimer(metrics.ReplicationTasksFetched, time.Duration(len(taskInfos)))
 
-	var (
-		oldestUnprocessedTaskTimestamp = t.timeSource.Now().UnixNano()
-		oldestUnprocessedTaskID        = t.ackLevels.GetTransferMaxReadLevel()
-		msgs                           = &types.ReplicationMessages{
-			// Happy path assumption - we will push all tasks to replication tasks.
-			ReplicationTasks:       make([]*types.ReplicationTask, 0, len(tasks)),
-			HasMore:                hasMore,
-			LastRetrievedMessageID: lastReadTaskID,
-		}
-	)
+	// Happy path assumption - we will push all tasks to replication tasks.
+	msgs := &types.ReplicationMessages{
+		ReplicationTasks:       make([]*types.ReplicationTask, 0, len(taskInfos)),
+		LastRetrievedMessageID: lastReadTaskID,
+		HasMore:                hasMore,
+	}
 
-	if len(tasks) > 0 {
+	if len(taskInfos) > 0 {
 		// it does not matter if we can process task or not, but we need to know what was the oldest task information we have read.
 		// tasks must be ordered by taskID/time.
-		oldestUnprocessedTaskID = tasks[0].TaskID
-		oldestUnprocessedTaskTimestamp = tasks[0].CreationTime
+		oldestUnprocessedTaskID = taskInfos[0].TaskID
+		oldestUnprocessedTaskTimestamp = taskInfos[0].CreationTime
 	}
 
-	for _, task := range tasks {
-		replicationTask, err := t.store.Get(ctx, pollingCluster, *task)
+	t.scope.RecordTimer(metrics.ReplicationTasksLagRaw, time.Duration(t.ackLevels.GetTransferMaxReadLevel()-oldestUnprocessedTaskID))
```

Comment:
See @Groxx's point about timers: We may be forced to migrate away from these fairly soon. Maybe that can be part of a different diff (I think you're just refactoring it, but just FYI)

Reply:
Yeah, I have just moved them. This is the reason why new metric `replication_tasks_batch_size` is a gauge metric.